vterebey/budgetcalculator/budget_calculator_class.py

```diff
@@ -0,0 +1,302 @@
+month in the format: YYYY-MM)
+5. Display expenses by category and month at the same time
+(see above how the category and month are set)
+All expenses are recorded in the file 'values.csv'
+The file contains the category,date and cost
+"""
+
+
+import csv
+import re
+from datetime import datetime
+import matplotlib.pyplot as plt
+
+
+# regular expressions for category validation
+fieldnames = ["Category", "Date", "Amount"]
+CAT_REGEX = "^[0-9]+$"
+cat_pattern = re.compile(CAT_REGEX)
+default_categories = {
+    1: "Grocery store",
+    2: "Clothing",
+    3: "Entertainment",
+    4: "Cafe",
+    5: "Store",
+    6: "Pharmacy",
+}
+with open("values.csv", "w", encoding="UTF-8", newline="") as file:
+    writer = csv.writer(file)
+    writer.writerow(fieldnames)
+
+
+class Budget:
+
+    "A class to represent a budget"
+
+    categories = {}
+
+    def __init__(self, category, date, amount):
+        """
+        This function initializes expense values
+        """
+        self.category = category
+        self.amount = amount
+        self.date = date
+
+    def _add_expense(self):
+        """
+        This function adds category and it's amount to the dictionary
+        for displaying statistics
+        """
+        if self.category not in Budget.categories:
+            Budget.categories[self.category] = self.amount
+        else:
+            Budget.categories[self.category] += self.amount
+
+    def expense_record(self):
+        """This function records expenses to a file"""
+        if int(self.amount) == self.amount:
+            self.amount = int(self.amount)
+        with open("values.csv", "a", encoding="UTF-8", newline="") as csvfile:
+            writer_ = csv.writer(csvfile)
+            writer_.writerow([self.category, self.date, self.amount])
+        Budget._add_expense(self)
+
+    def expenses_stat(self):
+        """This function dispays expenses in the form of a chart"""
+        amounts = list(Budget.categories.values())
+        # pylint: disable=W0612
+        fig1, ax1 = plt.subplots()
+        ax1.pie(
+            amounts,
+            labels=list(Budget.categories.keys()),
+            autopct="%1.1f%%",
+            shadow=False,
+            startangle=90,
+        )
+        ax1.axis("equal")
+        ax1.set_title("Statistics")
+        plt.show()
+
+    def group_by_cat(self, category):
+        """This function groups expenses by category"""
+        with open("values.csv", "r", encoding="UTF-8", newline="") as csvfile:
+            reader = csv.DictReader(csvfile, delimiter=",")
+            print(*fieldnames, sep=",")
+            for row in reader:
+                if row["Category"] == category:
+                    print(f"{row['Category']},{row['Date']},{row['Amount']}")
+        print()
+
+    def group_by_month(self, month):
+        """This function groups expenses by month"""
+        with open("values.csv", "r", encoding="UTF-8", newline="") as csvfile:
```

Comment:
if we use some constant text in the code, such as "values.csv", the it is better to create some constant like STORAGE_FILENAME to avoid cases when we need to update it in all places

Reply:
Updated